pkg/controller/build/build_controller.go

```diff
@@ -240,6 +241,13 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 	b.Status.Registered = corev1.ConditionFalse
 	b.Status.Reason = succeedStatus
 
+	// Validate if remote repository exists
+	if err := r.validateSourceUrl(ctx, b, b.Spec.Source.URL); err !=nil {
+		b.Status.Reason = err.Error()
+		updateErr := r.client.Status().Update(ctx, b)
+		return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
+	}
```

Comment:
@HeavyWombat If my understanding is correct. I still think we need to both return validateSourceUrl error and updateError. Because whatever validate failed or update build status failed, we both need to reconcile.

If we only return updateError, once update succeed, then it will return `nil`, then reconcile won't happen. However, validate source url indeed failed. I think this is not expected.

How do you think? Pls correct me if my understanding incorrect, thanks!

Reply:
@qu1queee 
>The rational is that if a spec.source.url is not valid, the only way for end-users to address this is by writing in their YAML the correct "url", and by doing this, end-users will trigger an UPDATE event in the Build leading to one more reconciliation. Based on this rational, we do not need to trigger reconciles whenever the spec.source.url validations fails, either the user fix this leading to the UPDATE or the Build remains in failure. Does this make sense?
>
Yes, I see. I misunderstand this before. Change it, thanks for your clarification.

Reply:
@xiujuan95 maybe Im missing something but for the use-case you are mentioning I dont think is a problem. If the sourceURL validation fails, you dont reconcile again. There is only one scenario where u might reconcile again, and that is if the update of the `Status.Reason` fails, and in this case if the secret is missing, then you just mark the Build as failed(one more time) and update the `Status.Reason` with a reference to the missing secret.

Reply:
@qu1queee Yes, in this case, build controller marks build as failed and update the `Status.Reason` is secret missing. But I think the first error `sourceURL validation failed` has been flooded. (This is the logic error of my code)

But I agree controller reconciles when validate failed maybe decrease our controller's performance. So we don't want to build controller reconciles when validation failed, only reconcile while update build status failed. So this makes sense for me.

I will change it to only return updateError. If update build status successfully, then I will tell controller stop reconciling until a UPDATE event happens, then `sourceURL validation failed has been flooded` problem will be fixed.

Reply:
There is one more thing missing here, we should not return the error of the validation as seen in:

```go
	// Validate if remote repository exists
	if err := r.validateSourceURL(ctx, b, b.Spec.Source.URL); err != nil {
		b.Status.Reason = err.Error()
		if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
			return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr) <-----
		}
		return reconcile.Result{}, nil
	}
```
the error of the validation is not the reason why we want to Reconcile again, but rather the error when trying to Update the Status, it should look like:

```go
	if err := r.validateSourceURL(ctx, b, b.Spec.Source.URL); err != nil {
		b.Status.Reason = err.Error()
		if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
			return reconcile.Result{}, updateErr
		}
		return reconcile.Result{}, nil
	}
```

Reply:
Done!

Reply:
thanks,looks good, resolving